src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateManager.HostStates.cs

```diff
@@ -64,73 +59,63 @@ private DiagnosticAnalyzerMap CreateLanguageSpecificAnalyzerMap(string language,
 
                 private class DiagnosticAnalyzerMap
                 {
-                    private readonly DiagnosticAnalyzer _compilerAnalyzer;
                     private readonly StateSet _compilerStateSet;
 
+                    private readonly ImmutableArray<StateSet> _orderedSet;
                     private readonly ImmutableDictionary<DiagnosticAnalyzer, StateSet> _map;
 
                     public DiagnosticAnalyzerMap(HostAnalyzerManager analyzerManager, string language, ImmutableDictionary<DiagnosticAnalyzer, StateSet> analyzerMap)
                     {
-                        // hold directly on to compiler analyzer
-                        _compilerAnalyzer = analyzerManager.GetCompilerDiagnosticAnalyzer(language);
+                        _map = analyzerMap;
+
+                        var compilerAnalyzer = analyzerManager.GetCompilerDiagnosticAnalyzer(language);
```

Comment:
I really don't like that we are tracking four different things now for analyzer prioritization.

* The compiler analyzers
* Implementations of `DocumentDiagnosticAnalyzer`
* Implementations of `ProjectDiagnosticAnalyzer`
* Everything else

Each of these four uses a different prioritization mechanism, and none of them leverage the functionality which users come to expect with MEF/VS (`OrderAttribute` / `ExtensionOrderAttribute`).

❓ What are our alternatives to unify and simplify this behavior?

Reply:
DiagnosticAnalyzer doesn't use MEF.

...

we have 4 different groups of analyzers (C#/VB compiler analyzer which is special, regular diagnostic analyzer, Document/ProjectDiagnosticAnalyzer, analyzer that implements IBuiltInAnalyzer). Engine treats them differently. reason is simple. since we want to allow special ability third party analyzer (regular diagnostic analyzer) can't use.

sure, we can have an issue tracking whether we should expose this special ability to third party or not. but that is not something this PR is about.